feat: реализовать Основной сервис

anzerdev opened the pull request.

Commit:
feat: реализовать Основной сервис

Commit:
feat: реализовать Основной сервис

Commit:
Merge remote-tracking branch 'origin/main_svc' into main_svc

# Conflicts:
#	main/pom.xml

Commit:
feat: реализовать Основной сервис

feat: реализовать Основной сервис

anzerdev committed:
Delete main/src/main/java/ru/practicum/ewm/main/event/model directory

anzerdev committed:
Delete statistic directory

alexache commented on service/pom.xml:
```diff
@@ -0,0 +1,172 @@
+        <dependency>
+            <groupId>org.mapstruct</groupId>
+            <artifactId>mapstruct</artifactId>
+            <version>1.5.3.Final</version>
```
Версии библиотек должны быть указаны только в родительском модуле

anzerdev replied:
исправлено

alexache commented on service/pom.xml:
```diff
@@ -0,0 +1,172 @@
+            <plugin>
+                <groupId>com.mysema.maven</groupId>
+                <artifactId>apt-maven-plugin</artifactId>
+                <version>1.1.3</version>
```
Как и версии плагинов - тоже в родительском модуле

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/category/dto/CategoryListDto.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+@Builder
+public class CategoryListDto {
```
Точно ли нужен такой класс? Можно ведь и просто List использовать как он есть

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/category/model/Category.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public int hashCode() {
+        return getClass().hashCode();
```
Хэшкод тоже лучше только по айди считать, чтобы ничего не сломать для особенных классов, как JPA модели

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/model/Event.java:
```diff
@@ -0,0 +1,60 @@
+    @Column(name = "VIEWS")
+    private Long views = 0L;
+    @OneToMany(mappedBy = "event")
+    private Set<Request> requests;
```
(вопрос) Действительно ли нужны эти поля в сервисе?

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/model/Event.java:
```diff
@@ -0,0 +1,60 @@
+    @Column(name = "TITLE", nullable = false)
+    private String title;
+    @Column(name = "CONFIRMED_REQUESTS")
+    private Integer confirmedRequests = 0;
```
Это не должно храниться в данной таблице - вместо этого нужно получать запросом в момент получения события (и избегать запросов в цикле)

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/model/Event.java:
```diff
@@ -0,0 +1,60 @@
+    @Column(name = "STATE")
+    private State state = State.PENDING;
+    @Column(name = "VIEWS")
+    private Long views = 0L;
```
Это тоже не должно храниться в базе, а запрашиваться в сервисе статистики

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -0,0 +1,357 @@
+                    if (request.getStatus().equals(RequestStatus.PENDING)) {
+                        if (event.getParticipantLimit() == 0) {
+                            request.setStatus(RequestStatus.CONFIRMED);
+                            event.setConfirmedRequests(event.getConfirmedRequests() + 1);
```
Как и писал - так делать не нужно. Нужно получить количество подтвержденных заявок перед проходом по списку, а затем отслеживать, если во время обработки этих запросов уже превысим лимит

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -0,0 +1,357 @@
+                    } else {
+                        rejectedRequests.add(requestDto);
+                    }
+                });
```
А где сохранение?

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -0,0 +1,357 @@
+                                                      LocalDateTime rangeStart, LocalDateTime rangeEnd,
+                                                      Boolean onlyAvailable, Pageable pageable,
+                                                      HttpServletRequest httpServletRequest) {
+        serviceStatsClient.postStatistic(httpServletRequest, "ewm-service");
```
отправка статистики есть, а использование?

anzerdev replied:
исправлено

alexache commented on stats/pom.xml:
```diff
@@ -0,0 +1,70 @@
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+	</properties>
+
+	<dependencyManagement>
```
Жто должно быть в родительском модуле

anzerdev replied:
исправлено

alexache commented on stats/pom.xml:
```diff
@@ -0,0 +1,70 @@
+		<dependency>
+			<groupId>org.apache.commons</groupId>
+			<artifactId>commons-lang3</artifactId>
+			<version>3.0</version>
```
Не должно быть версий здесь

anzerdev replied:
исправлено

alexache commented on stats/stats-server/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS hits CASCADE;
```
Сервис статистики должен был быть сделан на прошлом этапе, откуда столько новых правок?

anzerdev replied:
невнимательность, кривые руки и в связи с этим многократная помощь зала привели к этому((((

Commit:
fix: исправить замечания после ревью

Commit:
fix: исправить замечания после ревью

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -260,9 +254,12 @@ public ListEventShortDto getEventsByFiltersPublic(String text, List<Long> catego
+        List<EventShortDto> eventShortDtoList = eventMapper.mapToListEventShortDto(page.getContent());
+
+        eventShortDtoList.forEach(eventShortDto ->
+                eventShortDto.setViews(serviceStatsClient.getStatistic(eventShortDto.getId())));
```
Это запрос к сервису статистики в цикле, этого нужно избегать. Нужен один запрос по списку событий

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -246,8 +241,7 @@ public ListEventShortDto getEventsByFiltersPublic(String text, List<Long> catego
         if (onlyAvailable) {
-            booleanBuilder.and((QEvent.event.participantLimit.eq(0)))
-                    .or(QEvent.event.participantLimit.gt(QEvent.event.confirmedRequests));
+            booleanBuilder.and((QEvent.event.participantLimit.eq(0)));
```
А куда потеряли условие о том, что событие должно быть доступно для бронирования

anzerdev replied:
переписал немного метод

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -300,20 +297,20 @@ private BooleanBuilder createQuery(List<Long> ids, List<String> states, List<Lon
-                        } else if (event.getParticipantLimit() > event.getConfirmedRequests()) {
+                        } else if (event.getParticipantLimit() > eventConfirmedRequests) {
                             if (!event.getRequestModeration()) {
                                 request.setStatus(RequestStatus.CONFIRMED);
```
Если мы подтверждаем какую-то заявку, то количество доступных мест должно сокращаться?

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -329,8 +326,10 @@ private void moderateRequests(List<RequestDto> confirmedRequests,
                 .forEach(requestDto -> {
                     if (requestDto.getStatus().equals(RequestStatus.CONFIRMED)) {
                         confirmedRequests.add(requestDto);
+                        requestRepository.save(requestMapper.mapToRequest(requestDto));
```
Не стоит делать запросы к базе в цикле

anzerdev replied:
исправлено

Commit:
fix: исправить замечания после ревью2

fix: исправить замечания после ревью2

fix: исправить замечания после ревью2

fix: исправить замечания после ревью2

fix: исправить замечания после ревью2

Commit:
fix: исправить замечания после ревью2

Commit:
fix: исправить замечания после ревью2

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -254,12 +255,36 @@ public List<EventShortDto> getEventsByFiltersPublic(String text, List<Long> cate
+        List<Event> eventList = page.getContent();
+        List<EventShortDto> eventShortDtoList = eventMapper.mapToListEventShortDto(eventList).stream()
+                .peek(eventShortDto -> {
+                    eventShortDto.setViews(serviceStatsClient.getStatistic(eventShortDto.getId()));
```
Это все еще запрос к сервису статистики в цикле. Нужен один запрос по списку ивентов

anzerdev replied:
исправлено

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -305,18 +330,25 @@ private void moderateRequests(List<RequestDto> confirmedRequests,
                             if (!event.getRequestModeration()) {
                                 request.setStatus(RequestStatus.CONFIRMED);
+                                requestRepository.save(request);
+                                event.setParticipantLimit(event.getParticipantLimit() - 1);
```
Немного не это имел в виду - лимит на количество мест в событии не должен меняться. Нужно просто остлеживать количество уже подтвержденных заявок, то есть eventConfirmedRequests

anzerdev replied:
исправлено

Commit:
fix: исправить замечания после ревью3

Commit:
style: fix codestyle

alexache commented on service/src/main/java/ru/practicum/service/event/service/EventServiceImp.java:
```diff
@@ -271,6 +273,21 @@ public List<EventShortDto> getEventsByFiltersPublic(String text, List<Long> cate
         }
     }
 
+    private long getEventViews(Long eventId) {
```
Попытка исправления есть, но что поменялось? 
Этот метод так и вызывается в цикле - то есть на 20 событий 20 вызовов в сервис статистику. К тому же получение статистики всех возможных событий выглядит странно,
Как должно быть?

- Формируешь список событий
- получаешь для них URI
- вызываешь клиент статистики с этим списком URI  (API самого сервиса стастистики принимает список uris, клиент надо доработать)
- получаешь список объектов с просмотрами
- формируешь из этого списка мапу - eventId -> количество просмотров - обходя однократно список и вытаскивая айди
- в цикле по событиям подставляешь значения прсомотров из мапы

Commit:
fix: исправить замечания после ревью4
